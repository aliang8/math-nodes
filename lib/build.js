/**
 * Functions to build nodes
 */

const q = require('./query.js')

export const apply = (op, args, options = {}) => ({
    type: 'Apply',
    op: op,
    args: args,
    ...options,
})

// Operations

export const neg = (arg, options = {}) => apply('neg', [arg], options)      // options: wasMinus
export const add = (...terms) => apply('add', terms)
export const sub = (minuend, subtrahend) => apply('add', [minuend, neg(subtrahend, {wasMinus: true})])
export const mul = (...args) => apply('mul', args)
export const implicitMul = (...args) => apply('mul', args, {implicit: true})
export const div = (numerator, denominator) => apply('div', [numerator, denominator])
export const pow = (base, exponent) => apply('pow', [base, exponent])
export const abs = (arg) => apply('abs', [arg])
export const fact = (arg) => apply('fact', [arg])
export const nthRoot = (radicand, index) => apply('nthRoot', [radicand, index || number('2')])

// Relations

export const eq = (...args) => apply('eq', args)
export const ne = (...args) => apply('ne', args)
export const lt = (...args) => apply('lt', args)
export const le = (...args) => apply('le', args)
export const gt = (...args) => apply('gt', args)
export const ge = (...args) => apply('ge', args)


export const identifier = (name, options = {}) => ({
    type: 'Identifier',
    name: name,
    ...options, // options: subscript
})

export const number = (value, options = {}) => ({
    type: 'Number',
    value: String(value),
    ...options,
})


export const parens = (body, options = {}) => ({
    type: 'Parentheses',
    body: body,
    ...options,
})

// deprecated aliases
export const parensNode = parens
export const numberNode = number
export const identifierNode = identifier
export const applyNode = apply

// e.g 3 -> -3, -3x -> 3x, x + 3 -> -(x + 3), 2/3 -> -2 / 3
export const negate = (node) => {
    if (q.isNeg(node)) {
        return node.args[0]
    } else if (q.isAdd(node)) {
        return neg(node)
    } else if (q.isFraction(node)) {
        return div(negate(q.getNumerator(node)), q.getDenominator(node))
    } else if (q.isPolynomialTerm(node)) {
        if (q.isNeg(q.getCoefficient(node))) {
            return apply(
                'mul',
                [negate(q.getCoefficient(node)), ...node.args.slice(1)],
                {implicit: node.implicit})
        } else {
            return neg(node)
        }
    } 
}
