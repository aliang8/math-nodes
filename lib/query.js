/**
 * Functions to query properties of nodes
 */

const build = require('./build')
const {print} = require('math-parser')

export const isIdentifier = node => node && node.type === 'Identifier'
export const isApply = node => node && node.type === 'Apply'
export const isParens = node => node && node.type === 'Parentheses'

// deprecated, use isApply
export const isOperation = node => isApply(node) && !isNumber(node)
export const isFunction = node => isApply(node) && isIdentifier(node.op)

// TODO: curry it?
export const _isOp = (op, node) => isApply(node) && node.op === op

export const isAdd = node => _isOp('add', node)
export const isMul = node => _isOp('mul', node)
export const isDiv = node => _isOp('div', node)
export const isPow = node => _isOp('pow', node)
export const isNeg = node => _isOp('neg', node)
export const isPos = node => _isOp('pos', node)
export const isAbs = node => _isOp('abs', node)
export const isFact = node => _isOp('fact', node)
export const isNthRoot = node => _isOp('nthRoot', node)
export const isFraction = node => isNeg(node) ? isFraction(node.args[0]) : isDiv(node)
export const isConstantFraction = node => isFraction(node) && node.args.every(isNumber)
export const isIntegerFraction = node => isFraction(node) && node.args.every(isInteger)
export const isDecimal = node => isNumber(node) && getValue(node) % 1 != 0

const relationIdentifierMap = {
    'eq': '=',
    'lt': '<',
    'le': '<=',
    'gt': '>',
    'ge': '>=',
    'ne': '!=',
}

export const isRel = node => isApply(node) && node.op in relationIdentifierMap

export const isNumber = node => {
    if (node.type === 'Number') {
        return true
    } else if (isNeg(node)) {
        return isNumber(node.args[0])
    } else {
        return false
    }
}

export const isInteger = node => {
    return isNumber(node) && Number.isInteger(getValue(node))
}

export const isPolynomial = (node) => {
    if (isPolynomialTerm(node)) {
        return true
    } else if (isDiv(node)) {
        return isNumber(getDenominator(node)) && isPolynomial(getNumerator(node))
    } else if (isAdd(node) || isMul(node)) {
        return node.args.every(isPolynomialTerm)
    } else {
        return false
    }
}

export const isVariableFactor = (node) =>
    isIdentifier(node) ||
    isPow(node) && (isPolynomial(node.args[0]) || isPolynomialTerm(node))
    && (isNumber(node.args[1]) || isVariableFactor(node.args[1]))

export const isPolynomialTerm = (node) => {
    if (isNumber(node) || isConstantFraction(node) || isDecimal(node)) {
        return true
    } else if (isIdentifier(node)) {
        return true
    } else if (isPow(node)) {
        const [base, exponent] = node.args
        const positiveExp = isInteger(exponent) ? getValue(exponent) >= 1 : true
        return isPolynomial(base) && isPolynomial(exponent) && positiveExp
    } else if (isNeg(node)) {
        return isPolynomialTerm(node.args[0])
    } else if (isMul(node)) {
        return node.args.every(isPolynomialTerm) && node.implicit
    } else {
        return false
    }
}

export const isImplicitMul = (node) => {
    if (isMul(node)) {
        return node.implicit
    } else if (isNeg(node)) {
        return isImplicitMul(node.args[0])
    } else {
        return false
    }
}

// check if it's a number before trying to get its value
export const getValue = node => {
    if (node.type === 'Number') {
        return parseFloat(node.value)
    } else if (isNeg(node)) {
        return -getValue(node.args[0])
    } else if (isPos(node)) {
        return getValue(node.args[0])
    } else {
        return null
    }
}

export const getNumerator = (node) => {
    if (isFraction(node)) {
        return isNeg(node) ? getNumerator(node.args[0]) : node.args[0]
    } else {
        return null
    }
}

export const getDenominator = (node) => {
    if (isFraction(node)) {
        return isNeg(node) ? getDenominator(node.args[0]) : node.args[1]
    } else {
        return null
    }
}

export const hasConstantBase = node => isPow(node) &&
    (isNumber(node.args[0]) || isConstantFraction(node.args[0]) || isIntegerFraction(node.args[0]))

// TODO: handle multivariable polynomials
// Get degree of a polynomial term
// e.g. 6x^2 -> 2
export const getPolyDegree = (node) => {
    if (isNumber(node)) {
        return build.number(0)
    } else if (isIdentifier(node)) {
        return build.number(1)
    } else if (isPow(node)) {
        return node.args[1]
    } else if (isMul(node)){
        // TODO: Math.max(...node.args.map(getPolyDegree))
        // TODO: same base appearing more than once
        return getPolyDegree(node.args[1])
    } else if (isNeg(node)) {
        const variable = node.args[0]
        return getPolyDegree(variable.args[1])
    } else {
        return null
    }
}

// TODO: 2^2
export const getCoefficient = (node) => {
    if (isNumber(node) || hasConstantBase(node)) {
        return node
    } else if (isIdentifier(node) || isPow(node)) {
        return build.numberNode('1')
    } else if (isNeg(node)) {
        return build.neg(getCoefficient(node.args[0]), {wasMinus: node.wasMinus})
    } else if (isMul(node)) {
        const numbers = node.args.filter(arg => isNumber(arg) || isConstantFraction(arg))
        if (numbers.length > 1) {
            return build.mul(...numbers)
        } else if (numbers.length === 1) {
            return numbers[0]
        } else {
            return build.numberNode('1')
        }
    }
}

export const getVariableFactors = (node, factors = []) => {
    if (isVariableFactor(node)) {
        factors.push(node)
    } else if (isMul(node)) {
        node.args.forEach(arg => getVariableFactors(arg, factors))
    } else if (isNeg(node)) {
        // TODO: figure out how to handle (x)(-y)(z)
        return getVariableFactors(node.args[0])
    }
    return factors
}

export const getVariableFactorName = (node) => {
    if (isIdentifier(node)) {
        return node.name
    } else if (isPow(node)) {
        return getVariableFactorName(node.args[0])
    } else {
        return null
    }
}

export const sortVariables = (variables) =>
    variables.sort(
        (a, b) => getVariableFactorName(a) > getVariableFactorName(b))

/*
  Returns a dictionary containing a coefficient map, a list of constants
  and other miscellaenous terms
  The coefficient map is a dictionary with unique polynomial terms as keys
  and a list of respective coefficients that appear in the expression as values.

  e.g. 2x^2 + 5x^2 + 3x + 3
  {constants: [3], coefficientMap: ['x^2': [2, 5], 'x': [3]]}
*/
export const getCoefficientsAndConstants = (node, coefficientMap = {}, constants = [], others = []) => {
    if (isNumber(node) || isConstantFraction(node)) {
        constants.push(node)
    } else if (isFunction(node)) {
        // cos, sin, f(a), etc
        others.push(node)
    } else if (isPolynomialTerm(node) && !hasConstantBase(node)) {
        // x^2, 3x^2

        // sort the variables first (2yzx -> 2xyz)
        const sortedVariables = sortVariables(getVariableFactors(node))

        const coefficient = getCoefficient(node)
        const implicit = isImplicitMul(node)
        const key = sortedVariables.length > 1
            ? print(build.applyNode('mul', sortedVariables, {implicit}))
            : print(sortedVariables[0])

        if (!(key in coefficientMap)) {
            coefficientMap[key] = [coefficient]
        } else {
            coefficientMap[key].push(coefficient)
        }
    } else if(isPolynomial(node) || isApply(node)) {
        // 2x^2 + 3x + 1, 3x^2 * 2x^2
        node.args.forEach(function(arg) {
            getCoefficientsAndConstants(arg, coefficientMap, constants, others)
        })
    } else {
        others.push(node)
    }

    return {coefficientMap, constants, others}
}

// returns true if two nodes have the same base variable
// e.g. (x + 1)^1 and (x + 1)^3 , true
// e.g. x^2, (x + 1)^2, false
export const hasSameBase = (node1, node2) => {
    return nodeEquals(node1, node2)
        || isPow(node1) && nodeEquals(node1.args[0], node2)
        || isPow(node2) && nodeEquals(node1, node2.args[0])
        || isPow(node1) && isPow(node2) && nodeEquals(node1.args[0], node2.args[0])
}

export const nodeEquals = (node1, node2) => print(node1) === print(node2)
