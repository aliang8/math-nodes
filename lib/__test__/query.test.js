import assert from 'assert'
import {parse} from 'math-parser'
import stringify from 'json-stable-stringify'

import * as query from '../query'
import * as build from '../build'

const reverseAlphabetical = (a, b) => a.key < b.key ? 1 : -1

const serializer = {
    print(val) {
        return stringify(val, {cmp: reverseAlphabetical, space: '    '})
    },
    test() {
        return true
    },
}

expect.addSnapshotSerializer(serializer)  // eslint-disable-line

const suite = (name, func, cases) => {
    describe(name, () => {
        cases.forEach((c) => {
            it(c, () => {
                const result = func(parse(c))
                expect(result).toMatchSnapshot()  // eslint-disable-line
            })
        })
    })
}

suite.only = (name, func, cases) => {
    describe.only(name, () => {
        cases.forEach((c) => {
            it(c, () => {
                const result = func(parse(c))
                expect(result).toMatchSnapshot()  // eslint-disable-line
            })
        })
    })
}

describe('query', () => {
    let a, x, y

    beforeEach(() => {
        a = {type: 'Number', value: '5'}
        x = {type: 'Identifier', name: 'x'}
        y = {type: 'Identifier', name: 'y'}
    })

    suite('getVariableFactors', query.getVariableFactors, [
        '2x',
        'x^2',
        '2x^2',
        '2 x y z',
    ])

    suite('getCoefficientsAndConstants', query.getCoefficientsAndConstants, [
        '2',
        '2x',
        '2x^2',
        '2 x y z',
        '2x^3 + 3',
        'x^3 + y^3 + x y z + 3',
        '2/3 + 3 + cos(4)',
        '3x^2 * 2x^2',
        '2/3(x + 1)^1'
    ])

    it('isVariableFactor', () => {
        assert(query.isVariableFactor(parse('x')))
        assert(query.isVariableFactor(parse('xyz')))
        assert(query.isVariableFactor(parse('x^2')))
        assert(query.isVariableFactor(parse('(x+1)^2')))
        assert(!query.isVariableFactor(parse('2x^2')))
        assert(!query.isVariableFactor(parse('x y z')))
        assert(!query.isVariableFactor(parse('x + y')))
    })

    it('isPolynomialTerm', () => {
        assert(query.isPolynomialTerm(parse('x')))
        assert(query.isPolynomialTerm(parse('2x')))
        assert(query.isPolynomialTerm(parse('2x^2')))
        assert(query.isPolynomialTerm(parse('(x+2)^2')))
        assert(query.isPolynomialTerm(parse('(x+2)^(x+2)')))
        assert(query.isPolynomialTerm(parse('-2')))
        assert(query.isPolynomialTerm(parse('2.2')))
        assert(!query.isPolynomialTerm(parse('x^-2')))
    })

    it('isPolynomial', () => {
        // test cases from: https://www.mathsisfun.com/algebra/polynomials.html
        assert(query.isPolynomial(parse('1/5')))
        assert(query.isPolynomial(parse('5')))
        assert(query.isPolynomial(parse('3x')))
        assert(query.isPolynomial(parse('x/2')))
        assert(query.isPolynomial(parse('(3x)/8')))
        assert(query.isPolynomial(parse('x - 2')))
        assert(query.isPolynomial(parse('-6y^2 - (7/9)x')))
        assert(query.isPolynomial(parse('3xyz + 3xy2z - 0.1xz - 200y + 0.5')))
        assert(query.isPolynomial(parse('512v^5 + 99w^5')))
        assert(query.isPolynomial(parse('(2x^2 + 3y^2)/3')))
        assert(!query.isPolynomial(parse('3xy^-2')))
        assert(!query.isPolynomial(parse('2/(x+2)')))
        assert(!query.isPolynomial(parse('1/x')))
    })

    it('hasSameBase', () => {
        assert(query.hasSameBase(parse('x^1'), parse('x^2')))
        assert(query.hasSameBase(parse('(x+1)^2'), parse('(x+1)^x')))
        assert(query.hasSameBase(parse('(xyz)^2'), parse('(xyz)^5')))
        assert(!query.hasSameBase(parse('(x+1)^2'), parse('(x+2)^2')))
    })

    it('isIdentifier', () => {
        assert(query.isIdentifier(x))
        assert(!query.isIdentifier(a))
    })

    it('isApply', () => {
        assert(query.isApply({type: 'Apply'}))
        assert(!query.isApply(a))
        assert(!query.isApply(x))
    })

    it('isFunction', () => {
        assert(query.isFunction({
            type: 'Apply',
            op: {type: 'Identifier', name: 'sin'},
            args: [x],
        }))
        assert(!query.isFunction({type: 'Apply', op: 'sin'}))
    })

    it('isAdd', () => {
        assert(query.isAdd(parse('1 + 2')))
        assert(query.isAdd(parse('1 + 2 + x + y')))
        // This is b/c subtraction get's parsed as adding a negative
        assert(query.isAdd(parse('1 - 2')))
        assert(!query.isMul({
            type: 'Apply',
            op: 'add',
            args: build.number(1),
        }))
    })

    it('isMul', () => {
        assert(query.isMul(parse('2 * x')))
        assert(query.isMul(parse('2 x')))
        assert(!query.isMul({
            type: 'Apply',
            op: 'mul',
            args: build.number(1),
        }))
    })

    it('isDiv', () => {
        assert(query.isDiv(parse('x / y')))
        assert(!query.isDiv({
            type: 'Apply',
            op: 'div',
            args: [1, 2, 3].map(build.number),
        }))
    })

    it('isNeg', () => {
        assert(query.isNeg(parse('-x')))
        assert(query.isNeg(parse('-2')))
        // Even though this resolves to a positive 'x', we only check the given
        // node.
        assert(query.isNeg(parse('--x')))
        assert(!query.isDiv({
            type: 'Apply',
            op: 'neg',
            args: [1, 2].map(build.number),
        }))
    })

    it('isPos', () => {
        assert(query.isPos(parse('+x')))
        assert(query.isPos(parse('+2')))
        assert(query.isPos(parse('++x')))
        assert(!query.isDiv({
            type: 'Apply',
            op: 'pos',
            args: [1, 2].map(build.number),
        }))
    })

    it('isAbs', () => {
        assert(query.isAbs(parse('|x + y|')))
        assert(!query.isDiv({
            type: 'Apply',
            op: 'abs',
            args: [1, 2].map(build.number),
        }))
    })

    it('isFact', () => {
        assert(query.isFact(parse('x!')))
        assert(!query.isDiv({
            type: 'Apply',
            op: 'fact',
            args: [1, 2].map(build.number),
        }))
    })

    it('isNthRoot', () => {
        assert(query.isNthRoot(parse('nthRoot(4)')))
        assert(query.isNthRoot(parse('nthRoot(8, 3)')))
        assert(!query.isDiv({
            type: 'Apply',
            op: 'nthRoot',
            args: [1, 2, 3].map(build.number),
        }))
    })

    it('isFraction', () => {
        assert(query.isFraction(parse('2 / 3')))
        assert(query.isFraction(parse('-2 / 3')))
        assert(query.isFraction(parse('-(2 / 3)')))
        assert(query.isFraction(parse('--(2 / 3)')))
    })

    it('isConstantFraction', () => {
        assert(query.isConstantFraction(parse('2.2 / 3')))
        assert(!query.isConstantFraction(parse('x / 3')))
    })

    it('isIntegerFraction', () => {
        assert(!query.isIntegerFraction(parse('2.2 / 3')))
        assert(query.isIntegerFraction(parse('2 / 3')))
        assert(!query.isIntegerFraction(parse('x / 3')))
    })

    it('hasConstantBase', () => {
        assert(query.hasConstantBase(parse('2^2')))
        assert(query.hasConstantBase(parse('(2/3)^2')))
        assert(query.hasConstantBase(parse('(-2)^2')))
        assert(!query.hasConstantBase(parse('x^2')))
    })

    it('isDecimal', () => {
        assert(query.isDecimal(parse('-2.2')))
        assert(query.isDecimal(parse('2.2')))
        assert(!query.isDecimal(parse('-2')))
        assert(!query.isDecimal(parse('2')))
    })

    const relations = ['eq', 'ne', 'lt', 'le', 'gt', 'ge']

    it('isRel', () => {
        relations.forEach(rel => {
            assert(query.isRel({
                type: 'Apply',
                op: rel,
                args: [x, y]
            }))
        })
        assert(!query.isRel(a))
        assert(!query.isRel(x))
    })

    it('isNumber', () => {
        assert(query.isNumber(a))
        assert(query.isNumber(parse('-2')))
        assert(!query.isNumber(x))
    })

    it('getValue', () => {
        assert.equal(query.getValue(parse('2')), 2)
        assert.equal(query.getValue(parse('-2')), -2)
        assert.equal(query.getValue(parse('2.2')), 2.2)
        assert.equal(query.getValue(parse('-2.2')), -2.2)
        assert.equal(query.getValue(x), null)
    })

    it('getNumerator', () => {
        assert.deepEqual(query.getNumerator(parse('2/3')), parse('2'))
        assert.deepEqual(query.getNumerator(parse('-2/3')), parse('-2'))
        assert.deepEqual(query.getNumerator(parse('-(2/3)')), parse('2'))
        assert.equal(query.getNumerator(x), null)
    })

    it('getDenominator', () => {
        assert.deepEqual(query.getDenominator(parse('2/3')), parse('3'))
        assert.deepEqual(query.getDenominator(parse('2/-3')), parse('-3'))
        assert.deepEqual(query.getDenominator(parse('-(2/3)')), parse('3'))
        assert.equal(query.getDenominator(x), null)
    })

    it('getPolyDegree', () => {
        assert.deepEqual(query.getPolyDegree(parse('2')), parse('0'))
        assert.deepEqual(query.getPolyDegree(parse('2x')), parse('1'))
        assert.deepEqual(query.getPolyDegree(parse('2xyz')), parse('1'))
        assert.deepEqual(query.getPolyDegree(parse('2x^2')), parse('2'))
        assert.deepEqual(query.getPolyDegree(parse('x^2')), parse('2'))
        assert.deepEqual(query.getPolyDegree(parse('-2x^2')), parse('2'))
        assert.deepEqual(query.getPolyDegree(parse('-2/3 x^2')), parse('2'))
    })

    it('getCoefficient', () => {
        assert.deepEqual(query.getCoefficient(parse('2')), parse('2'))
        assert.deepEqual(query.getCoefficient(parse('2^2')), parse('2^2'))
        assert.deepEqual(query.getCoefficient(parse('2^(2/3)')), parse('2^(2/3)'))
        assert.deepEqual(query.getCoefficient(parse('2x')), parse('2'))
        assert.deepEqual(query.getCoefficient(parse('2x^2')), parse('2'))
        assert.deepEqual(query.getCoefficient(parse('x^2')), parse('1'))
        assert.deepEqual(query.getCoefficient(parse('-2x^2')), parse('-2'))
        assert.deepEqual(query.getCoefficient(parse('2/3 x^2')), parse('2/3'))
        assert.deepEqual(query.getCoefficient(parse('-2/3 x^2')), parse('-2/3'))
    })
})
