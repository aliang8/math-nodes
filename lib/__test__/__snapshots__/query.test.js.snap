// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`query getCoefficientsAndConstants 2 1`] = `
{
    "others": [
    ],
    "constants": [
        {
            "value": "2",
            "type": "Number"
        }
    ],
    "coefficientMap": {
    }
}
`;

exports[`query getCoefficientsAndConstants 2 x y z 1`] = `
{
    "others": [
    ],
    "constants": [
    ],
    "coefficientMap": {
        "x y z": [
            {
                "value": "2",
                "type": "Number"
            }
        ]
    }
}
`;

exports[`query getCoefficientsAndConstants 2/3 + 3 + cos(4) 1`] = `
{
    "others": [
        {
            "type": "Apply",
            "op": {
                "type": "Identifier",
                "name": "cos"
            },
            "args": [
                {
                    "value": "4",
                    "type": "Number"
                }
            ]
        }
    ],
    "constants": [
        {
            "type": "Apply",
            "op": "div",
            "args": [
                {
                    "value": "2",
                    "type": "Number"
                },
                {
                    "value": "3",
                    "type": "Number"
                }
            ]
        },
        {
            "value": "3",
            "type": "Number"
        }
    ],
    "coefficientMap": {
    }
}
`;

exports[`query getCoefficientsAndConstants 2/3(x + 1)^1 1`] = `
{
    "others": [
    ],
    "constants": [
    ],
    "coefficientMap": {
        "(x + 1)^1": [
            {
                "type": "Apply",
                "op": "div",
                "args": [
                    {
                        "value": "2",
                        "type": "Number"
                    },
                    {
                        "value": "3",
                        "type": "Number"
                    }
                ]
            }
        ]
    }
}
`;

exports[`query getCoefficientsAndConstants 2x 1`] = `
{
    "others": [
    ],
    "constants": [
    ],
    "coefficientMap": {
        "x": [
            {
                "value": "2",
                "type": "Number"
            }
        ]
    }
}
`;

exports[`query getCoefficientsAndConstants 2x^2 1`] = `
{
    "others": [
    ],
    "constants": [
    ],
    "coefficientMap": {
        "x^2": [
            {
                "value": "2",
                "type": "Number"
            }
        ]
    }
}
`;

exports[`query getCoefficientsAndConstants 2x^3 + 3 1`] = `
{
    "others": [
    ],
    "constants": [
        {
            "value": "3",
            "type": "Number"
        }
    ],
    "coefficientMap": {
        "x^3": [
            {
                "value": "2",
                "type": "Number"
            }
        ]
    }
}
`;

exports[`query getCoefficientsAndConstants 3x^2 * 2x^2 1`] = `
{
    "others": [
    ],
    "constants": [
    ],
    "coefficientMap": {
        "x^2": [
            {
                "value": "3",
                "type": "Number"
            },
            {
                "value": "2",
                "type": "Number"
            }
        ]
    }
}
`;

exports[`query getCoefficientsAndConstants x^3 + y^3 + x y z + 3 1`] = `
{
    "others": [
    ],
    "constants": [
        {
            "value": "3",
            "type": "Number"
        }
    ],
    "coefficientMap": {
        "y^3": [
            {
                "value": "1",
                "type": "Number"
            }
        ],
        "x^3": [
            {
                "value": "1",
                "type": "Number"
            }
        ],
        "x y z": [
            {
                "value": "1",
                "type": "Number"
            }
        ]
    }
}
`;

exports[`query getVariableFactors 2 x y z 1`] = `
[
    {
        "type": "Identifier",
        "name": "x"
    },
    {
        "type": "Identifier",
        "name": "y"
    },
    {
        "type": "Identifier",
        "name": "z"
    }
]
`;

exports[`query getVariableFactors 2x 1`] = `
[
    {
        "type": "Identifier",
        "name": "x"
    }
]
`;

exports[`query getVariableFactors 2x^2 * y 1`] = `
[
    {
        "type": "Apply",
        "op": "pow",
        "args": [
            {
                "type": "Identifier",
                "name": "x"
            },
            {
                "value": "2",
                "type": "Number"
            }
        ]
    },
    {
        "type": "Identifier",
        "name": "y"
    }
]
`;

exports[`query getVariableFactors 2x^2 1`] = `
[
    {
        "type": "Apply",
        "op": "pow",
        "args": [
            {
                "type": "Identifier",
                "name": "x"
            },
            {
                "value": "2",
                "type": "Number"
            }
        ]
    }
]
`;

exports[`query getVariableFactors x^2 * 2/3 y^2 1`] = `
[
    {
        "type": "Apply",
        "op": "pow",
        "args": [
            {
                "type": "Identifier",
                "name": "x"
            },
            {
                "value": "2",
                "type": "Number"
            }
        ]
    },
    {
        "type": "Apply",
        "op": "pow",
        "args": [
            {
                "type": "Identifier",
                "name": "y"
            },
            {
                "value": "2",
                "type": "Number"
            }
        ]
    }
]
`;

exports[`query getVariableFactors x^2 1`] = `
[
    {
        "type": "Apply",
        "op": "pow",
        "args": [
            {
                "type": "Identifier",
                "name": "x"
            },
            {
                "value": "2",
                "type": "Number"
            }
        ]
    }
]
`;
